### PR TITLE
Reading in e2e data

### PR DESCRIPTION
**Pull request by evgeniiaraz**
**Proposed changes**:
- adding `E2EImporter` Tobias created;
- adding processing of e2e data from yaml and markdown

**Comments**:
- There are a lot of tests which fail in here but this is mostly because of NLU errors; 
All the tests are running properly in 'Changes to NLU ...' PR.
- I left the messages in `ActionExecuted` and `UserUttered` for now. I want to use them in later steps for some sanity checks of Core Featurization. Once that is done, I'll remove them

This PR is for clarity of how pipeline is changing 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by Ghostvv** [comment]
@evgeniiaraz could you please fix code quality failures

**Comment by evgeniiaraz** [comment]
TODO by vova's comment: add an end-to-end example in `md`

File: rasa/importers/importer.py [hunk 1]
```diff
@@ -271,3 +274,118 @@ async def get_nlu_data(self, language: Optional[Text] = "en") -> TrainingData:
...
+
+
+class E2EImporter(TrainingDataImporter):
+    """Importer which
+    - enhances the NLU training data with actions / user messages from the stories.
+    - adds potential end-to-end bot messages from stories as actions to the domain
+    """
+
+    def __init__(self, importer: TrainingDataImporter) -> None:
+        self._importer = importer
```
Comment (Ghostvv) [inline rasa/importers/importer.py:286]: why importer has `_importer` attribute?

File: rasa/nlu/constants.py [hunk 2]
```diff
@@ -1,9 +1,13 @@
 TEXT = "text"
 INTENT = "intent"
 RESPONSE = "response"
+ACTION_TEXT = "action_text"
 
 RESPONSE_KEY_ATTRIBUTE = "response_key"
 
+MESSAGE_ACTION_NAME = "action_name"
+MESSAGE_INTENT_NAME = "intent_name"
```
Comment (Ghostvv) [inline rasa/nlu/constants.py:9]: to be consistent with the other constants:
```suggestion
ACTION_NAME = "action_name"
INTENT_NAME = "intent_name"
```

File: rasa/core/training/story_reader/markdown_story_reader.py [hunk 3]
```diff
@@ -29,11 +29,14 @@ class MarkdownStoryReader(StoryReader):
 
     """
 
-    async def read_from_file(self, filename: Text,) -> List[StoryStep]:
+    async def read_from_file(self, filename: Text) -> List[StoryStep]:
         """Given a md file reads the contained stories."""
 
         try:
-            with open(filename, "r", encoding=io_utils.DEFAULT_ENCODING) as f:
+            # TODO: Hack by Genie for Windows
```
Comment (Ghostvv) [inline rasa/core/training/story_reader/markdown_story_reader.py:36]: why does it work currently for Windows?

File: data/test_yaml_stories/stories_hybrid_e2e.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+stories:
+- story: My hybrid End-to-End story
+  steps:
+  # Regular story with labels
+  - intent: simple
+  - action: utter_greet
+  # Actual messages are given instead of labels
+  - user: "I am looking for a [Kenyan](cuisine) restaurant"
```
[the remark this answers is not in the document]
Comment (wochinge) [reply data/test_yaml_stories/stories_hybrid_e2e.yml:8]: No that's correct. https://www.notion.so/rasa/Training-data-format-80c786f1996f491dbda0e095debc85fe#ce3805a9e791470dbfa9ef84f02457b9

File: data/test_yaml_stories/stories_hybrid_e2e.yml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  - action: utter_greet
+  # Actual messages are given instead of labels
+  - user: "I am looking for a [Kenyan](cuisine) restaurant"
+  - bot: "good for you"
+  # Regular labeled events
+  - intent: goodbye
+  - action: utter_goodbye
+  # Actual messages are given instead of labels
+  - user: One more thing
+  - bot: What?
```
Comment (wochinge) [inline data/test_yaml_stories/stories_hybrid_e2e.yml:15]:
```suggestion
  - bot: What?

```

File: rasa/importers/importer.py (hunk 1 above)
Comment (wochinge) [reply rasa/importers/importer.py:286]: Because we are using the [composite pattern](https://en.wikipedia.org/wiki/Composite_pattern) here. This code was actually already reviewed by Rick (https://github.com/RasaHQ/rasa/pull/6083)

File: rasa/core/events/__init__.py [hunk 6]
```diff
@@ -989,11 +996,17 @@ def __init__(
         confidence: Optional[float] = None,
         timestamp: Optional[float] = None,
         metadata: Optional[Dict] = None,
-    ):
+        # TODO: Throw out message attribute
+        message: Optional[Message] = None,
```
Comment (wochinge) [inline rasa/core/events/__init__.py:1000]: @evgeniiaraz This todo was just for the `e2e-wip-parser` branch. We don't need to carry it over.

File: rasa/core/training/story_reader/markdown_story_reader.py (hunk 3 above)
Comment (wochinge) [reply rasa/core/training/story_reader/markdown_story_reader.py:36]: @evgeniiaraz The MD stuff can be undone as it was just a temporary hack and we just need yaml

File: rasa/core/events/__init__.py (hunk 6 above)
Comment (evgeniiaraz) [reply rasa/core/events/__init__.py:1000]: I wanted to leave it in until I add the new core featurization for some sanity checks there and then remove it.

File: rasa/importers/importer.py [hunk 7]
```diff
@@ -271,3 +274,118 @@ async def get_nlu_data(self, language: Optional[Text] = "en") -> TrainingData:
...
+            # Simple cache to avoid loading all of this multiple times
+            self._cached_stories = await self._importer.get_stories(
+                interpreter, template_variables, use_e2e, exclusion_percentage
+            )
+        return self._cached_stories
+
+    async def get_config(self) -> Dict:
+        return await self._importer.get_config()
+
+    async def get_nlu_data(self, language: Optional[Text] = "en", only_nlu: bool = False) -> TrainingData:
```
Comment (wochinge) [inline rasa/importers/importer.py:331]: you can't just add a variable to an interface implementation @evgeniiaraz Also we need tests for this. And what's the use case for `only_nlu`?

**Comment by wochinge** [comment]
> I left the messages in ActionExecuted and UserUttered for now. I want to use them in later steps for some sanity checks of Core Featurization. Once that is done, I'll remove them

But this just transfers the current problem of the to huge PR to  the `e2e` branch, no?

File: rasa/core/training/story_reader/markdown_story_reader.py (hunk 3 above)
Comment (Ghostvv) [reply rasa/core/training/story_reader/markdown_story_reader.py:36]: let's keep for now

File: rasa/core/events/__init__.py (hunk 6 above)
Comment (Ghostvv) [reply rasa/core/events/__init__.py:1000]: let's treat these PRs as PRs against master, therefore there is no reason to introduce smth to remove it later

File: rasa/importers/importer.py (hunk 7 above)
Comment (evgeniiaraz) [reply rasa/importers/importer.py:331]: the use case is when the user does `rasa train nlu` and does not want any of default actions added.

Comment (wochinge) [reply rasa/importers/importer.py:331]: Mhm, I see. In that case it will use the `NluImporter` which means we have to change the `Domain.empty` call there and then we don't need the flag 🤔

Comment (evgeniiaraz) [reply rasa/importers/importer.py:331]: Sorry I forgot another part before -- it has to also ignore `_additional_training_data_from_stories()`.

File: rasa/nlu/constants.py (hunk 2 above)
Comment (evgeniiaraz) [reply rasa/nlu/constants.py:9]: Done in the next PR

Comment (Ghostvv) [reply rasa/nlu/constants.py:9]: why in the next PR? let's not mix PRs

File: rasa/core/training/story_reader/markdown_story_reader.py (hunk 3 above)
Comment (wochinge) [reply rasa/core/training/story_reader/markdown_story_reader.py:36]: @Ghostvv even if keep markdown, the e2e stuff shouldn't make into master as the format is completely off charts

File: rasa/importers/importer.py [hunk 8]
```diff
@@ -309,6 +310,8 @@ async def _get_domain_with_e2e_actions(self) -> Domain:
                 if isinstance(event, ActionExecuted) and event.e2e_text
             ]
 
+        additional_e2e_action_names = list(set(additional_e2e_action_names))
```
Comment (wochinge) [inline rasa/importers/importer.py:313]: let's treat it as set from the beginning and convert to `list` in the end instead of converting back and forth.
Please also add / update a test for the case of duplicate actions (sorry that I missed that)

File: rasa/importers/importer.py [hunk 9]
```diff
@@ -372,7 +376,14 @@ def _messages_from_user_utterance(event: UserUttered) -> Message:
 
 
 def _messages_from_action(event: ActionExecuted) -> Message:
-    return Message(event.e2e_text or "", data={MESSAGE_ACTION_NAME: event.action_name})
+    # we need to store the action text twice to be able to differentiate between user and bot text in NLU processing
```
Comment (wochinge) [inline rasa/importers/importer.py:379]: could you explain why? I think it should work without storing it twice as you don't add anymore information?

File: rasa/core/training/story_reader/markdown_story_reader.py (hunk 3 above)
Comment (Ghostvv) [reply rasa/core/training/story_reader/markdown_story_reader.py:36]: what do you mean by `format is completely off charts`?

File: rasa/importers/importer.py (hunk 8 above)
Comment (evgeniiaraz) [reply rasa/importers/importer.py]: Done both!

File: rasa/importers/importer.py (hunk 9 above)
Comment (evgeniiaraz) [reply rasa/importers/importer.py:381]: This is changed [here](https://github.com/RasaHQ/rasa/blob/d3ed2e84f80bb3a967c2f4edd39e432ee6420f7c/rasa/importers/importer.py#L380-L384). We make sure that user text and action text are stored separately.

**Comment by Ghostvv** [comment]
this PR will be part of https://github.com/RasaHQ/rasa/pull/6204#event-3589379133
